syft/execution/type_wrapper.py

```diff
@@ -0,0 +1,261 @@
+from syft_proto.execution.v1.type_wrapper_pb2 import ClassType as ClassTypePB
+
+class NestedTypeWrapper:
+    """
+        Class for input type serialization and type checking for nested structures.
+    """
+    def __init__(self, nested_type=None):
+        if nested_type:
+            self.serialized_nested_type = NestedTypeWrapper.serialize_nested_type(nested_type)
+        else:
+            self.serialized_nested_type = None
+
+    @staticmethod
+    def serialize_nested_type(input_arg: any) -> Union[list, tuple, dict, type]:
+        """
+            Function to serialize input of a function/Plan, including nested types.
+
+            Note: supported nested structures: list, tuple, dict with string keys.
+
+            Params:
+                input_arg: *args of a function or Plan.
+
+            Returns:
+                Union[list, tuple, dict, type]: Nested structured with types instead of objects.
+        """
+        if isinstance(input_arg, (list, tuple)):
+            result = [NestedTypeWrapper.serialize_nested_type(elem) for elem in input_arg]
+            return tuple(result) if isinstance(input_arg, tuple) else result
+
+        if isinstance(input_arg, dict):
+            serialized_dict = {k: NestedTypeWrapper.serialize_nested_type(v) for k, v in input_arg.items()}
+            return serialized_dict
+
+        return type(input_arg)
+
+    @staticmethod
+    def raise_typecheck_warn(obj_type: str, obj_name: str, build: str, call: str, path: str) -> None:
+        """
+            Function to raise a typecheck warning if two types differ.
+
+            Params:
+                obj_type: the type of the object returned by calling .__name__ on it.
+                obj_name: the name/id of the object.
+                build: the build/reference argument type.
+                call: the called argument type.
+                path: the nested path to reach that obj.
+
+            Returns:
+                 None
+        """
+        warn(
+            f"{obj_type} {obj_name} {path} has type {build}, while being built with type {call}.",
+            RuntimeWarning,
+        )
+
+    @staticmethod
+    def raise_missmatch_err(obj_type: str, obj_name: str, build: int, call: int, path: str) -> None:
+        """
+            Function to raise an error if two nested structures differ in length.
+
+            Params:
+                obj_type: the type of the object returned by calling .__name__ on it.
+                obj_name: the name/id of the object.
+                build: the build/reference argument length.
+                call: the called argument length.
+                path: the nested path to reach that obj.
+
+            Returns:
+                 None
+        """
+        raise TypeError(f"{obj_type} {obj_name} {path} has length {call}, while being build with length {build}.")
+
+    @staticmethod
+    def raise_wrong_no_arguments_err(obj_type: str, obj_name: str, build: int, call: int) -> None:
```

Comment:
I think no_arguments is a bit ambiguous, maybe nb_arguments or number_arguments would be better